enhance sweep to raster

egouden opened the pull request:
* link variable to spatial_ref for gdal compatiblity
* follow GIS convention for raster bounds order
* ensure raster precision with bounds and resolution
* option to snap bounds/resolution when creating raster
* drop gdal and rasterio dependency
* preserve original variable names

kmuehlbauer reviewed (commented):
@egouden I really appreciate your work on extending wradlib's functionality and also to make user experience better.

But for some reason I do not understand some of your design decisions. Maybe you can add some review comments here and there. Anyway, I really want wradlib to behave consistently for xarray objects. So those functionality should be aligned with existing code.

kmuehlbauer commented:
@egouden I've merged #714. You can build upon main now, adding your markdown text only notebook into the structure. You'll find the already converted markdown notebooks in https://github.com/wradlib/wradlib/tree/8c2ae6f341d39afd724aaa382e6e8d542c90e226/examples/notebooks.

egouden commented on wradlib/comp.py:
```diff
@@ -359,10 +370,9 @@ def wrapper(obj, **kwargs):
 
-    if isinstance(sweep, DataArray):
-        out.name = f"{sweep.name}.to_raster"
+    georef.add_raster_grid_mapping(out)
```
done

egouden commented:
I believe keyword arguments should be used for improving clarity and reducing maintenance costs. This is a significant issue which can be discussed separately. I removed the stars so you can focus on the review of the enhancements.

I still have to integrate the notebook.

egouden commented:
I have to do this locally:

(.venv) PS C:\Users\egoud\wradlib> python -m jupytext --to notebook --execute .\examples\notebooks\visualisation\plot_ppi.md --set-kernel python3

egouden commented:
I could not quickly find the reason for the error when running the test "without GDAL".

kmuehlbauer commented:
> I cannot find the reason for the error when running the test "without GDAL".

Sorry Edouard, I've pushed a fix. I've mixed up some imports.

kmuehlbauer commented:
@egouden I've checked the notebooks in the rendered docs. Somehow the geographic plots are missing the data.

Also, you would have to add the notebooks in the toc inside overview.md

kmuehlbauer commented:
This is the fix needed for no-gdal, just in case of overwriting my commit:

```
Subject: [PATCH] fix has_import
---
Index: wradlib/georef/projection.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/wradlib/georef/projection.py b/wradlib/georef/projection.py
--- a/wradlib/georef/projection.py	(revision 685caadecd0b52ecbd607f6bb42708104e932781)
+++ b/wradlib/georef/projection.py	(revision 466e99c0fb6cedfea45272bc8c982eaa00ef7795)
@@ -99,10 +99,10 @@
     if isinstance(crs, pyproj.CRS) and type(crs) is pyproj.CRS:
         if trg == "pyproj":
             return crs
-    elif has_import("cartopy") and isinstance(crs, cartopy.crs.CRS):
+    elif has_import(cartopy) and isinstance(crs, cartopy.crs.CRS):
         if trg == "cartopy":
             return crs
-    elif has_import("osr") and isinstance(crs, osr.SpatialReference):
+    elif has_import(osr) and isinstance(crs, osr.SpatialReference):
         if trg == "osr":
             return crs
         crs = crs.ExportToWkt()

```

egouden commented:
The earth projection with arc-second was not right. I fixed it.

I also added unit tests to verify gdal and rioxarray compatibility (as advertised in the feature).

Commit:
enhance sweep to raster

* link variable to spatial_ref for gdal compatiblity
* follow GIS convention for raster bounds order
* ensure raster precision with bounds and resolution
* option to snap bounds/resolution when creating raster
* drop gdal and rasterio dependency
* preserve original variable names

kmuehlbauer commented:
@egouden Is this now ready for review? For the future, please do not force-push! Instead add commits to your PR and if need be merge wradlib main into your PR. Otherwise comments placed at certain locations during review are disconnected and reviewers can't follow the changes you made during lifetime of your PR.

I'll try to review this today, to get this in soon.

egouden commented:
@kmuehlbauer 

Thank you for your collaboration on this PR.  I am now marking each significant change with a commit.

I cannot find explicit guidance in the documentation. Force-push (amend) is actually promoted in the short guide that I introduced recently. The PR management is a fundamental and often neglected issue in my opinion.

The main advantage of amend is to keep a clean history with an updated commit message, which can be further reviewed. I think this is a good strategy for complex PR with many changes.

I see a benefit of several commits (one per reviewer comment) for simple PR. But this is costly for complex PR.

I am curious to learn about your experience with other packages.

Also the use of draft should be encouraged to inform the reviewer about the status.

Commit:
FIX notebook temp file name

egouden commented:
> > I cannot find explicit guidance in the documentation. Force-push (amend) is actually promoted in the short guide that I introduced recently. The PR management is a fundamental and often neglected issue in my opinion.
> 
> We might need to refine the short-guide or make it a bit more explicit in that regard.
> 
> Overall this is an issue with GitHub UI, see:
> 
> * https://github.com/orgs/community/discussions/3478
> * https://github.com/orgs/community/discussions/142466
> 
> But it is also has general implication:
> 
> * https://matthew-b-groves.medium.com/matts-tidbits-23-why-to-avoid-force-pushing-691d3abbab48
> 
> I've secured wradlib's branches to not accept force-pushes by accident. Contributors can develop in their branches as needed.
> 
> > The main advantage of amend is to keep a clean history with an updated commit message, which can be further reviewed. I think this is a good strategy for complex PR with many changes.
> 
> It does have an impact on the review process. For `xarray` it's good practice, that maintainers can push commits to PR branches to assist, same with doing this via suggestions UI. If a maintainer pulls a PR branch and works with it locally, rebasing or force-pushing to that branch will lead to conflicts.
> 
> > I see a benefit of several commits (one per reviewer comment) for simple PR. But this is costly for complex PR.
> 
> There does not need to be one commit per different change when introducing the PR, but when the PR is live, we should maintain it's history to reflect the changes during the lifetime of the PR.
> 
> > I am curious to learn about your experience with other packages regarding PR efficiency.
> 
> What I got used to is the approach I adapted to when working with `xarray`. The PR branch in general will not be rebased or force-pushed until review process is finished. And even then no manual rebase/force-push is needed as there will only be one final commit to the main branch (squash-merge). This keeps the history in main very clean with one commit per PR.
> 
> > Also the use of draft should be encouraged to inform the reviewer about the status.
> 
> I agree, if a PR is in draft mode, it should not be expected that it receives a review. So rebase/force-pushing in that stage should be allowed for easier work when developing.

I suggest the following strategy then:

- developer force-push in draft mode, address or reply reviewer comments (if not in starting phase), rebase systematically, make improvements if relevant
- when ready developer ask for review, new commit but with final message, reviewer checks all the changes 

I am not sure this is handled properly in github UI but at least can be done locally.

kmuehlbauer commented on wradlib/georef/projection.py:
```diff
@@ -849,32 +861,33 @@ def project_bounds(bounds, crs):
-    one_degree_longitude_meters = one_degree_latitude_meters * np.cos(lat_mid)
-    xsize = xsize / one_degree_longitude_meters
-    xsize = np.abs(xsize)
+    R = 6371000  # Earth radius in meters
```
Can we parametrize earth radius this as kwarg?

kmuehlbauer commented on wradlib/comp.py:
```diff
@@ -380,9 +390,9 @@ def togrid(self, *args, **kwargs):
     def sweep_to_raster(self, *args, **kwargs):
         if not isinstance(self, CompMethods):
-            return sweep_to_raster(self, *args, **kwargs)
+            return sweep_to_raster(sweep=self, **kwargs)
```
This needs fixing (keyword argument usage)?

kmuehlbauer reviewed (changes requested):
Hi @egouden, I've worked through his and have yet some more questions and suggestions. In general this looks like a really nice addition.

egouden commented on wradlib/georef/raster.py:
```diff
@@ -726,89 +813,142 @@ def create_raster_xarray(crs, bounds, size):
 
-def create_raster_geographic(bounds, size, size_in_meters=False):
-    """Create a geographic raster.
+def create_raster_geographic(
```
I have spent quite some time investigating this issue before submitting.

The main idea is to avoid precision issues indeed. Choosing 0.1 degree as resolution for example will raise an error in create_raster_xarray because it has no binary representation and cannot exactly divide the extent.

I could not find an easy way to deal with precision issues related to decimal degrees. One might allow representable decimal degree like 0.125. 

Note that conversion to degrees might be needed when exporting to other file format (e.g. tiff) for maximum compatibility.

kmuehlbauer replied:
OK, please add a comment in the code somewhere (and maybe also in the docstring). Just that we do not forget that. Maybe we can iterate later finding a proper solution. Thanks!

egouden commented on wradlib/georef/raster.py:
```diff
@@ -726,89 +813,142 @@ def create_raster_xarray(crs, bounds, size):
+    return edges
+
+
+def get_raster_coordinates(ds: xr.Dataset, mode: str = "center") -> np.ndarray:
```
OK done

Commit:
enhance sweep to raster

    * link variable to spatial_ref for gdal compatiblity
    * follow GIS convention for raster bounds order
    * ensure raster precision with bounds and resolution
    * option to snap bounds/resolution when creating raster
    * drop gdal and rasterio dependency
    * preserve original variable names

egouden commented:
This is ready for review. The comments have been addressed. A few minor improvements have been introduced.

kmuehlbauer commented on wradlib/comp.py:
```diff
@@ -286,28 +291,45 @@ def transform_binned(sweep, raster):
+        ) from err
+
+    radar_crs = sweep.xradar.get_crs()
+    crs_wkt = raster.spatial_ref.attrs["crs_wkt"]
```
Are we confident, that there will always be a `spatial_ref` variable? No blocker, but we should be prepared that users might hit an obstacle here.

kmuehlbauer commented on wradlib/georef/raster.py:
```diff
@@ -712,12 +716,93 @@ def raster_to_polyvert(dataset):
+        target (int): Desired resolution (size of each cell), must be a positive integer.
+
+    Returns:
+        int: Snapped resolution that evenly divides the extent and is closest to the target.
```
```suggestion
    Parameters
    ----------
    extent : int
        Total extent along one axis (e.g., width or height in arcseconds or pixels).
    target : int
        Desired resolution (size of each cell), must be a positive integer.

    Returns
    -------
    resolution : int
        Snapped resolution that evenly divides the extent and is closest to the target.
```

kmuehlbauer commented on wradlib/georef/raster.py:
```diff
@@ -712,12 +716,93 @@ def raster_to_polyvert(dataset):
+        resolution (int or tuple[int, int]): Desired resolution per axis.
+
+    Returns:
+        tuple[int, int, int, int]: Snapped bounds aligned to the resolution grid.
```
```suggestion
    Parameters
    ----------
    bounds :  tuple[int, int, int, int]
        (minx, miny, maxx, maxy) in integer units.
    resolution :  int or tuple[int, int])
        Desired resolution per axis.

    Returns
    -------
    bounds : tuple[int, int, int, int]
        Snapped bounds aligned to the resolution grid.
```

kmuehlbauer commented on wradlib/georef/raster.py:
```diff
@@ -726,89 +811,173 @@ def create_raster_xarray(crs, bounds, size):
+    Returns
+    -------
+    xr.Dataset
+        xarray dataset with WGS84 CRS and coordinates using arcsecond unit, following CF conventions.
```
```suggestion
    raster : xarray:Dataset
        xarray dataset with WGS84 CRS and coordinates using arcsecond unit, following CF conventions.
```

kmuehlbauer reviewed (changes requested):
Some minor docstring issues. Two occasions where keywords are still used. I've added a couple of suggestions, please double check.

After fixing this is ready to get in.

egouden commented on wradlib/georef/projection.py:
```diff
@@ -849,32 +859,35 @@ def project_bounds(bounds, crs):
+        longitude (float): Reference longitude in degrees.
+
+    Returns:
+        tuple: (delta_latitude, delta_longitude) in degrees
```
Thanks @kmuehlbauer for the review. I agree with all suggested changes. I assume it is ready on your side.  Would you like to take it from here?

kmuehlbauer replied:
@egouden Yes, I can commit the suggestions right away.

kmuehlbauer committed:
Apply suggestions from code review

kmuehlbauer committed:
fix arguments

kmuehlbauer committed:
fix more docstrings

kmuehlbauer committed:
fix more docstrings

kmuehlbauer reviewed (approved):
Thanks very much Edouard!
